xgcm/grid_ufunc.py

```diff
@@ -336,30 +360,106 @@ def apply_as_grid_ufunc(
             boundary_width=boundary_width_real_axes,
             boundary=boundary,
             fill_value=fill_value,
         )
+    else:
+        boundary_width_real_axes = {
+            real_ax: (0, 0) for real_ax in dummy_to_real_axes_mapping.values()
+        }
+        padded_args = args
+
+    if (
+        any(
+            _has_chunked_core_dims(arg, core_dims)
+            for arg, core_dims in zip(args, in_core_dims)
+        )
+        and map_overlap
+    ):
+        # map operation over dask chunks along core dimensions
+        from dask.array import map_overlap
+
+        # merge any lonely chunks from padding
+        rechunked_padded_args = _rechunk_to_merge_in_boundary_chunks(
+            padded_args,
+            args,
+            boundary_width_real_axes,
+            grid,
+        )
+
+        print(boundary_width_real_axes)
+        # TODO should this be adjusted?
+        boundary_width_per_numpy_axis = {
+            grid.axes[ax_name]._get_axis_dim_num(args[0]): width
+            for ax_name, width in boundary_width_real_axes.items()
+        }
+
+        def _dict_to_numbered_axes(sizes):
+            return tuple(sizes[dim] for dim in sizes.keys())
+
+        output_chunk_length_adjustment = _check_if_length_will_change(  # noqa
+            in_dummy_ax_names, out_dummy_ax_names, in_ax_pos, out_ax_pos
+        )
+
+        original_chunksizes = [arg.chunksizes for arg in args]
+        if output_chunk_length_adjustment == 0:
+            # Output chunks are the same as input chunks
+            # TODO first argument only because map_overlap can't handle multiple return values (I think)
+            true_chunksizes = original_chunksizes[0]
+            true_chunksizes_per_numpy_axis = _dict_to_numbered_axes(true_chunksizes)
+        else:
+            # the true chunks will differ from the original chunks if the ufunc changes the array's chunking or size
+            print(f"output chunk adjustment = {output_chunk_length_adjustment}")
+
+            # TODO currently only allowed to have one chunked core dim
+            core_dim_which_is_changing_size = dummy_to_real_axes_mapping[
+                in_dummy_ax_names[0][0]
+            ]
+            print(core_dim_which_is_changing_size)
+
+            print(original_chunksizes)
+            # create the "true" output chunks
+            # TODO first argument only because map_overlap can't handle multiple return values (I think)
+            true_chunksizes = original_chunksizes.copy()[0]
+            # TODO how should this operation work? Which specific chunk should it adjust the length of?
+            # (as is it wouldn't work because we're adding an int to a tuple)
+            true_chunksizes[core_dim_which_is_changing_size] = (
+                true_chunksizes[core_dim_which_is_changing_size]
+                + output_chunk_length_adjustment
+            )
+            true_chunksizes_per_numpy_axis = _dict_to_numbered_axes(true_chunksizes)
```

Comment:
@rabernat which specific chunk should I be changing the size of here?

Reply:
It's meant to be equivalent to the logic in cell 19 of [your notebook](https://gist.github.com/rabernat/0135a7b03b46705da7877147a3736aa8), i.e.

```python
orig_chunks = data_padded_rechunked.chunks
core_dim_chunks = orig_chunks[-1]
true_core_dim_chunks = ((core_dim_chunks[0]-1,) + core_dim_chunks[1:],)
true_chunks = orig_chunks[:-1] + true_core_dim_chunks

data_diffed = dsa.map_overlap(
    diff,
    data_padded_rechunked,
    depth={1: (1, 0)},
    boundary='none',
    trim=False,
    meta=np.array([], dtype=data_padded_rechunked.dtype),
    chunks=true_chunks
)
data_diffed
```

Reply:
If memory serves, this "chunk hacking" is directly tied to `boundary='none'` and `depth={1: (1, 0)},`. By specifying `trim=False`, we tell `map_overlap` that our function will take care of the trimming. This works great for all of the chunks that do not touch the global boundary, which are padded before being passed to our function `diff`. However, at this point, we have applied the global array boundary condition manually (creating a "padded" array), thus growing the array by one point in a particular direction (left, right, or both). The applied function (here `diff`) also trims those boundary chunks, but Dask is not expecting that to happen (since we specified `boundary=none`), so without `chunks=true_chunks`, it guesses the wrong chunk shape. (Try removing that option and see what happens.)

So there is a symmetry between what you pass to `depth` and how you set `true_chunks`. The key `1` tells you that we need to modify the expected chunks on axis 1. The value `(1, 0)` tells you that we need to shrink the first chunk along that axis by 1 element. `(0, 1)` would mean shrink the last chunk.

Make sense?

Reply:
This discussion is made redundant by https://github.com/xgcm/xgcm/pull/380/commits/070cddd70b46bd76f01191f31bad966320aec2e6